### PR TITLE
closes #264, check for system wide installation of geth

### LINKED ISSUE #264: Look for a system wide installation of geth or eth
I'm maintaining the mist packages for ArchLinux and need some guidance.

Starting up mist always tries to start a geth node in `${pkgdir}/usr/share/${pkgname}/resources/node/geth/geth`, but usually people install and maintain their own version of geth or eth in their system, e.g., `/usr/bin/geth`...

Would it be possible that mist checks whether any suitable client is available in `$PATH` before starting it's own node?

Discussion from [AUR](https://aur.archlinux.org/packages/mist/):

> For some odd reason /usr/share/mist/resources/node/geth/geth has no execution privileges after installation with pacman. This is why I'm getting the same error as @notungx when I run mist. Adding the privilege solves the problem.
> 
> The file is not marked as executable probably because it's unpacked from a .zip achive, which has no concept of access privileges. I see that PKGBUILD explicitly sets privileges for libnode.so and Ethereum-Wallet. It should be doing that to geth as well. Please add this line at the end of package():
> 
> ```
> chmod 755 "${pkgdir}/usr/share/${pkgname}/resources/node/geth/geth"
> ```
> 
> For a moment I thought that maybe lack of exacutable flag is deliberate - after all geth is an optional dependency here. But no - mist tries to use its own copy of geth even if you install the geth package from AUR. It does not seem to detect /usr/bin/geth. The dependency is misleading and should be removed.
> 
> The error and the dependency make it look as if geth wasn't bundled with mist until very recently. Is this what happened here? 

And:

> I get the following error after executing mist:
> 
> ```
> CONNECT to IPC PATH: /home/santi/.ethereum/geth.ipc
> CONNECT to IPC PATH: /home/santi/.ethereum/geth.ipc
> NODECONNECTOR ERROR { [Error: connect ECONNREFUSED /home/santi/.ethereum/geth.ipc]
> code: 'ECONNREFUSED',
> errno: 'ECONNREFUSED',
> syscall: 'connect',
> address: '/home/santi/.ethereum/geth.ipc' }
> Node type: geth
> Network: main
> Start node from /usr/share/mist/resources/node/geth/geth
> Stopping nodes...
> Starting geth node...
> UNCAUGHT EXCEPTION { [Error: spawn EACCES] code: 'EACCES', errno: 'EACCES', syscall: 'spawn' }
> Stopping nodes...
> ```
> 
> In my laptop works just executing mist without the need of executing geth first... any ideas?

Should I make the shipped geth binary executable for now? I would prefer using the system wide installation of geth.

### PR DESCRIPTION
**Pull request by luclu**
This current state of this PR will make `getNodePaths.js - getNodePaths()` to 
 - check for nodes in `./nodes/<type>/<platform>/<type>`
 - check if system-installation of that node is available (Linux + OS X)
 - gather and compares their versions
 - stores the relevant (latest version's) paths in a `resolvedPaths[type]` dictionary

Ideally this should be run early during startup as it can take up to 2 seconds to resolve the versions for (geth, eth and parity installed).

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ var filenameLowercase = 'mist';
 var filenameUppercase = 'Mist';
 var applicationName = 'Mist'; 
 var electronVersion = '1.2.5';
-var gethVersion = '1.4.10';
+module.exports.gethVersion = '1.4.10';
```
Comment (luclu) [inline gulpfile.js]: This feels ugly + travis doesn't like this:
```
/home/travis/build/ethereum/mist/gulpfile.js:213

        if(!!~file.path.indexOf('-'+ gethVersion +'-')) {

                                     ^
```

File: modules/getNodePath.js [hunk 2]
```diff
@@ -25,15 +27,44 @@ module.exports = function(type) {
...
+                
+                var systemGethVersion;
+                setTimeout(function () {
+                    if(systemGethPath !== null) {
+                        log.error(systemGethPath);
+                        var proc = exec(systemGethPath + ' version', 
+                            function(error, stdout, stderr) {
+                                systemGethVersion = stdout.match(/(?:[\d.]+)/)[0];
+                                log.warn('Geth node version: ' + systemGethVersion);
+                                log.warn(gulpfile.gethVersion);
```
Comment (luclu) [inline modules/getNodePath.js]: What is a better way to pull the internal version of geth?
 - creating an additional local variable?
 - pulling the version with `geth version`, too?

File: modules/getNodePath.js [hunk 3]
```diff
@@ -25,15 +27,44 @@ module.exports = function(type) {
...
+                var proc = exec('which geth', 
+                    function(error, stdout, stderr) {
+                        systemGethPath = stdout.toString().replace(/\n/, '');
+                        if (error !== null) {
+                                log.error("error exec 'which geth': " + error + stdout);
+                        }
+                    });
+                
+                var systemGethVersion;
+                setTimeout(function () {
```
Comment (luclu) [inline modules/getNodePath.js]: I currently use `child_process.exec` to execute the two commands `which geth` and `geth version` and use a timeout to wait for the async call to finish. 
What better approaches are there? Especially considering the later program flow depends on its output?

File: gulpfile.js (hunk 1 above)
Comment (hiddentao) [inline gulpfile.js]: My fix for the Linux node download issue will replace this with something better.

File: modules/getNodePath.js [hunk 4]
```diff
@@ -25,15 +27,44 @@ module.exports = function(type) {
...
+
     if (globallySetType) {
         resolvedPaths[type] = globallySetType;
     } else {
         var binPath = (Settings.inProductionMode)
             ? binaryPath.replace('nodes','node') + '/'+ type +'/'+ type
             : binaryPath + '/'+ type +'/'+ process.platform +'-'+ process.arch + '/'+ type;
 
-        if(Settings.inProductionMode) {
+        if(/*Settings.inProductionMode*/ true) {
```
Comment (hiddentao) [inline modules/getNodePath.js]: ?? Perhaps we need to get rid of the condition altogether then?

File: modules/getNodePath.js (hunk 3 above)
Comment (hiddentao) [inline modules/getNodePath.js]: I would recommend against using `which`. The `type` command is probably what we need, see http://superuser.com/questions/351889/what-is-the-unix-command-to-find-out-what-executable-file-corresponds-to-a-given

File: modules/getNodePath.js (hunk 2 above)
Comment (hiddentao) [inline modules/getNodePath.js]: Using `geth version` for the internal geth exec sounds good to me. This is also a way of verifying that it executes properly on the local OS.

File: modules/getNodePath.js (hunk 4 above)
Comment (luclu) [inline modules/getNodePath.js]: Yeah this is just for testing.
Sounds good to me.

File: modules/getNodePath.js (hunk 3 above)
Comment (luclu) [inline modules/getNodePath.js]: Ah great, I actually run into the [non-existing exit code issue](https://stackoverflow.com/questions/592620/check-if-a-program-exists-from-a-bash-script).

**Comment by luclu** [comment]
This current state of this PR will make `getNodePaths.js - getNodePaths()` to 
 - check for nodes in `./nodes/<type>/<platform>/<type>`
 - check if system-installation of that node is available
 - gather and compares their versions
 - stores the relevant (latest version's) paths in a `resolvedPaths[type]` dictionary

Ideally this should be run early during startup as it can take up to 2 seconds to resolve the versions for (geth, eth and parity installed).

**Comment by luclu** [comment]
argh...

**Pull request closed without merging**